flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java

```diff
@@ -73,7 +77,7 @@ public void open(MetricConfig config) {
 	@Override
 	public void report() {
 		try {
-			pushGateway.push(CollectorRegistry.defaultRegistry, jobName);
+			pushGateway.push(CollectorRegistry.defaultRegistry, jobName, instance);
```

Comment:
I would just re-use the `jobName`. This is effectively already an instance label.

Reply:
> In Prometheus terms, an endpoint you can scrape is called an instance, usually corresponding to a single process. A collection of instances with the same purpose, a process replicated for scalability or reliability for example, is called a job.

> For example, an API server job with four replicated instances:
job: api-server
--  instance 1: 1.2.3.4:5670
--  instance 2: 1.2.3.4:5671
--  instance 3: 5.6.7.8:5670
--  instance 4: 5.6.7.8:5671

https://prometheus.io/docs/concepts/jobs_instances/#jobs-and-instances

I think a Flink job corresponds to a Prometheus job, and taskmanager and jobmanager correspond to different instances. If the jobName is used as the instance label, the same metrics of different tasksmanages will conflict, and operations such as sum will fail.

I failed to use ip:port to form an instance label,  because I am not familiar enough with the Flink source. So  use this the easiest way

Reply:
No, a Flink job does not correspond to a Prometheus job. This doesn't work for session clusters (i.e. running multiple jobs), or for clusters that are just idling. Even if it did you wouldn't want an arbitrary random ID (you _never_ want random data) but something like the job ID.

As of right now we expect the jobName to be unique. It is thus safe to use as the `instance` as well.
Actually, this begs the rather interesting question whether you've set the jobName to some constant across multiple containers (as of right now, you shouldn't!)

Alternatively the instance label should also be made configurable, in a similar vein as the job label. In hindsight we made a mistake and should've made the `jobName` configuration apply to `instance` instead; the more you know.

One could argue that `jobName` could be something like `dispatcher` or `taskmanager`, with `instance` being the corresponding ID, but unfortunately not all IDs are properly exposed (see 
FLINK-9543)

Reply:
First of all, thank you very much for the information provided, which will help me better understand Flink related knowledge.

This question depends on how we view the Flink cluster.

First, consider the cluster as a black box. JM is the same as "dispatch" in SpringMVC. That whole cluster is a Promethues Job.

Second, the JM class is compared to a more feature-rich nginx. Then different jobs correspond to different Promethues Jobs.

Third, treat each Flink Job as a microservice cluster. The same function of Task and even Sub-Task is a specific micro-service function, a series of associated micro-services complete the entire business. Because the above two methods need to solve a problem, different TM runs this different Flink Job's task, the same task runs on different TMs. This means that the TM's logo cannot accurately separate the tasks of different Flink Jobs, metrics of the same name still conflict . JM and TM are just the running containers for jobs. So in theory, we need to be precise to the Task and even the Sub-Task level, then use the Job prefix to group them.

To solve this problem perfectly, we need a comprehensive understanding of the entire Metric system. Maybe we should close this PR and decide how to deal with it after a thorough discussion in the community.

Reply:
We had a similar discussion back when we added the pushgateway reporter.

The gist is that 2)/3) are not desirable since it makes the internals of the reporter rather complex (additionally for 2) the question remains what to do with clusters that aren't running a job) and 1) requires additional work outside of the metric system first to actually identify clusters. As of right now there's no concept of a cluster ID _anywhere_, and it doesn't make sense to only introduce it for the Prometheus reporter specifically. The current implementation for `jobName` was very much just a compromise.

The thing is the following: Ultimately a user can already group metrics in any way he/she desires. The jobName/instance stuff is, for our purposes, exclusively used to prevent conflicts between different containers. That's it, there's no semantic value in them since they (will, after FLINK-9543) only duplicate the container id label that we attach anyway for consistency with other reporters.
As such it isn't necessary to introduce additional complexity or significant reporter-specific logic.

Reply:
In fact, using a random string is also a compromise. However, if you only use **jobName** as the instance, putgateway will merge into one according to the **metric name and instance**, and the data will be lost. Just like the picture I wrote in JIRA  (https://issues.apache.org/jira/browse/FLINK-11742 ). Now I adjust the instance to a splicing of jobName and random strings, which can reduce confusion.

Reply:
Compared with **instance = ""**, **instance = jobname** just enriches the meaning of the indicator, but does not solve the data loss problem.

Reply:
How did you configure the jobName when you had the data loss? Was it a unique value for each container?

Reply:
Just like the official website said。
In addition, I tried to distinguish the job manager and task manager's jobName to solve the problem of empty tasks, but failed.

> metrics.reporter.promgateway.class: org.apache.flink.metrics.prometheus.PrometheusPushGatewayReporter
metrics.reporter.promgateway.host: localhost
metrics.reporter.promgateway.port: 9091
metrics.reporter.promgateway.jobName: myJob
metrics.reporter.promgateway.randomJobNameSuffix: false
metrics.reporter.promgateway.deleteOnShutdown: true

Reply:
Does it work if `randomJobNameSuffix` is enabled?

Reply:
> Does it work if `randomJobNameSuffix` is enabled?

In that case, the job names of each jobManger and taskmanger are different, and the instance division becomes unnecessary.

I looked at the metrics sent to pushgateway and thought that **`metrics.reporter.promgateway.jobName`** was a compromise on the Job concept in Prometheus. In fact, we can see that any metric associated with a specific Flink Job has a job_name label, which is our real "**`jobName`**". The xxx..jobName in the configuration file is more like the name of a cluster. Therefore, it is necessary to keep the `names of a cluster ` 
 consistent, and  more convenient to monitor.

Reply:
I get that, but I'm very much against introducing the concept of a "cluster_id" in a single reporter.

There is value in having it, but I'd much rather have a a) solution that works for all reporters and b) has value outside the metric system.
Note that this isn't really a metric issue as such an ID should also be visible (and realistically contained) within each JM/TM/RM/Dispatcher who forward this to the metric system as they do with other IDs.

When that is in place we can discuss whether we rework the jobName/instance stuff. jobName could be the cluster ID, and instance the ID of the JM/TM/etc. . However my feeling is that such a change would be unnecessary as the only requirement is that _some_ label allows filtering by cluster ID, not necessarily the `instance` one.